File: DSA/Python/src/dsa/lib/algo/sort/insertion_sort.py
from typing import List


class InsertionSort:
    def sort(self, ary: List[int]) -> None:
        n: int = len(ary)
        for i in range(1, n):
            b: int = ary[i]
            pos: int = self.insert_pos_of(ary, 0, i)
            ary[pos] = b

    @staticmethod
    def insert_pos_of(ary: List[int], st: int, idx: int) -> int:
        """
        Find the insert position of an element in the sorted array.

        :param ary:
        :param st: The start position of the sorted array.
        :param idx: The index of the element to be inserted. The element should be next to the end element of
        the sorted array
        :return: The index of insert position
        """
        b: int = ary[idx]
        idx -= 1
        while idx >= st:
            a = ary[idx]
            if a > b:
                ary[idx + 1] = a
            else:
                break
            idx -= 1
        return idx + 1

File: DSA/Java/src/dsa/lib/algo/sort/InsertionSort.java
package dsa.lib.algo.sort;

public class InsertionSort {
    public static void sort(int[] ary) {
        for (int i = 1; i < ary.length; ++i) {
            int b = ary[i];
            final int pos = insertPosOf(ary, 0, i);
            ary[pos] = b;
        }
    }

    /**
     * Find the insert position for an element in the sorted array
     *
     * @param ary
     * @param st  The start of the sorted array, inclusive.
     * @param idx The index of element to be inserted. The element should be next to the end of the sorted array.
     * @return The index of insert position
     */
    private static int insertPosOf(int[] ary, int st, int idx) {
        int b = ary[idx];
        for (--idx; idx >= st; --idx) {
            final int a = ary[idx];
            if (a > b) ary[idx + 1] = a;
            else break;
        }
        return idx + 1;
    }
}
